File: motor3d 2.0.ConvertedToC#/Recta3D.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using Motor3D.Algebra;
using System.Math;

namespace Motor3D.Espacio3D
{
	public class Recta3D : ObjetoGeometrico3D
	{

		private Plano3D PlanoA;
		private Plano3D PlanoB;
		private Vector3D mVector;
		private Punto3D mPunto;

		private Punto3D mPuntoMira;
		public Plano3D PrimerPlano {
			get { return PlanoA; }
		}

		public Plano3D SegundoPlano {
			get { return PlanoB; }
		}

		public Vector3D VectorDirector {
			get { return mVector; }
		}

		public Punto3D PuntoInicial {
			get { return mPunto; }
		}

		public Punto3D PuntoMira {
			get { return mPuntoMira; }
		}

		public Matriz[] Matrices {
			get { return RepresentacionMatricial(this); }
		}

		public Punto3D ObtenerPuntoParametrico(double Delta)
		{
			return new Punto3D(mPunto.X + (Delta * mVector.X), mPunto.Y + (Delta * mVector.Y), mPunto.Z + (Delta * mVector.Z));
		}

		public double ObtenerParametro(double Valor, EnumEjes Coordenada)
		{
			switch (Coordenada) {
				case EnumEjes.EjeX:
					if (mVector.X != 0) {
						return (Valor - mPunto.X) / mVector.X;
					} else {
						throw new ExcepcionGeometrica3D("RECTA3D (OBTENERPARAMETRO): Division por cero." + Constants.vbNewLine + "Coordenada: " + Coordenada.ToString() + Constants.vbNewLine + "Vector direccion: " + mVector.ToString());
					}
					break;
				case EnumEjes.EjeY:
					if (mVector.Y != 0) {
						return (Valor - mPunto.Y) / mVector.Y;
					} else {
						throw new ExcepcionGeometrica3D("RECTA3D (OBTENERPARAMETRO): Division por cero." + Constants.vbNewLine + "Coordenada: " + Coordenada.ToString() + Constants.vbNewLine + "Vector direccion: " + mVector.ToString());
					}
					break;
				case EnumEjes.EjeZ:
					if (mVector.Z != 0) {
						return (Valor - mPunto.Z) / mVector.Z;
					} else {
						throw new ExcepcionGeometrica3D("RECTA3D (OBTENERPARAMETRO): Division por cero." + Constants.vbNewLine + "Coordenada: " + Coordenada.ToString() + Constants.vbNewLine + "Vector direccion: " + mVector.ToString());
					}
					break;
			}
		}

		public Punto3D ObtenerPunto(double Valor, EnumEjes Coordenada)
		{
			return ObtenerPuntoParametrico(ObtenerParametro(Valor, Coordenada));
		}

		public Punto3D ObtenerPunto(double X)
		{
			Ecuacion ec1 = null;
			Ecuacion ec2 = null;
			SistemaEcuaciones sis = null;

			ec1 = new Ecuacion(PlanoA.B, PlanoA.C, -(PlanoA.D + (PlanoA.A * X)));
			ec2 = new Ecuacion(PlanoB.B, PlanoB.C, -(PlanoB.D + (PlanoB.A * X)));

			sis = new SistemaEcuaciones(ec1, ec2);

			if (sis.Solucion.TipoSolucion == TipoSolucionSistema.SistemaCompatibleDeterminado) {
				return new Punto3D(X, sis.Solucion.ValorSolucion[0], sis.Solucion.ValorSolucion[1]);
			} else {
				throw new ExcepcionGeometrica3D("RECTA3D (OBTENERPUNTO): No se ha podido calcular el punto. Es posible que los datos de los planos sean erroneos, o que el cálculo del sistema halla fallado." + Constants.vbNewLine + "Valor de la variable=" + X.ToString() + Constants.vbNewLine + "Primer plano: " + PlanoA.ToString() + Constants.vbNewLine + "Seundo plano: " + PlanoB.ToString() + Constants.vbNewLine + "Primera ecuación del sistema: " + ec1.ToString() + Constants.vbNewLine + "Segunda ecuación del sistema: " + ec2.ToString() + Constants.vbNewLine + "Solución obtenida: " + sis.Solucion.ToString());
			}
		}

		public Recta3D(Plano3D P1, Plano3D P2)
		{
			PlanoA = P1;
			PlanoB = P2;
			mVector = (P1.VectorNormal + P2.VectorNormal).VectorUnitario;
			mPunto = ObtenerPuntoParametrico(0);
			mPuntoMira = ObtenerPuntoParametrico(10);
		}

		public Recta3D(Punto3D P1, Punto3D P2) : this(P1, new Vector3D(P1, P2))
		{
		}

		public Recta3D(params Matriz[] Matrices)
		{
			if (Matrices.GetUpperBound(0) == 1) {
				Punto3D Punto = new Punto3D(Matrices[0]);
				Punto3D Punto2 = new Punto3D(Matrices[1]);

				mPunto = Punto;
				mPuntoMira = Punto2;
				mVector = new Vector3D(mPunto, mPuntoMira).VectorUnitario;

				//ÉSTOS VALORES SE OBTIENEN AL DESPEJAR LA ECUACIÓN PARAMÉTRICA DE UNA RECTA GENÉRICA:
				PlanoA = new Plano3D(mVector.Y, -mVector.X, 0, (-mVector.Y * Punto.X) + (mVector.X * Punto.Y));
				PlanoB = new Plano3D(0, mVector.Z, -mVector.Y, (-mVector.Z * Punto.Y) + (mVector.Y * Punto.Z));
			} else {
				throw new ExcepcionGeometrica3D("RECTA3D (NEW): La representación matricial de una recta corresponde a un array con dos matrices" + Constants.vbNewLine + "Tamaño del array=" + Matrices.GetUpperBound(0) + 1);
			}
		}

		public Recta3D(Punto3D Punto, Vector3D Vector)
		{
			mPunto = Punto;
			mPuntoMira = new Punto3D(mPunto.X + Vector.X, mPunto.Y + Vector.Y, mPunto.Z + Vector.Z);
			mVector = Vector.VectorUnitario;

			//ÉSTOS VALORES SE OBTIENEN AL DESPEJAR LA ECUACIÓN PARAMÉTRICA DE UNA RECTA GENÉRICA:
			PlanoA = new Plano3D(Vector.Y, -Vector.X, 0, (-Vector.Y * Punto.X) + (Vector.X * Punto.Y));
			PlanoB = new Plano3D(0, Vector.Z, -Vector.Y, (-Vector.Z * Punto.Y) + (Vector.Y * Punto.Z));
		}

		public bool Pertenece(Punto3D Punto)
		{
			return PlanoA.Pertenece(Punto) && PlanoB.Pertenece(Punto);
		}

		public override string ToString()
		{
			return "{Recta " + PlanoA.ToString() + " & " + PlanoB.ToString() + "}";
		}

		public static Matriz[] RepresentacionMatricial(Recta3D Recta)
		{
			Matriz[] Retorno = new Matriz[2];

			Retorno[0] = Recta.PuntoInicial.Matriz;
			Retorno[1] = Recta.PuntoMira.Matriz;

			return Retorno;
		}

		public static PosicionRelativa3D PosicionRelativa(Recta3D R1, Recta3D R2)
		{
			SistemaEcuaciones Sistema = new SistemaEcuaciones(new Ecuacion(R1.VectorDirector.X, R2.VectorDirector.X, R2.PuntoInicial.X - R1.PuntoInicial.Y), new Ecuacion(R1.VectorDirector.Y, R2.VectorDirector.Y, R2.PuntoInicial.Y - R1.PuntoInicial.Y), new Ecuacion(R1.VectorDirector.Z, R2.VectorDirector.Z, R2.PuntoInicial.Z - R1.PuntoInicial.Z));

			switch (Sistema.Solucion.TipoSolucion) {
				case TipoSolucionSistema.SistemaCompatibleDeterminado:
					return new PosicionRelativa3D(new Punto3D(Sistema.Solucion.ValorSolucion[0], Sistema.Solucion.ValorSolucion[1], Sistema.Solucion.ValorSolucion[2]));
				case TipoSolucionSistema.SistemaCompatibleIndeterminado:
					return new PosicionRelativa3D(TipoPosicionRelativa3D.Coincidente);
				case TipoSolucionSistema.SistemaIncompatible:
					if (Sistema.RangoMatrizPrincipal == 1) {
						return new PosicionRelativa3D(TipoPosicionRelativa3D.Paralelo);
					} else {
						return new PosicionRelativa3D(TipoPosicionRelativa3D.Cruce);
					}
					break;
			}
		}

		public static double Distancia(Recta3D Recta, Punto3D Punto)
		{
			return (new Vector3D(Recta.ObtenerPuntoParametrico(0), Punto) + Recta.VectorDirector).Modulo;
		}

		public static double Distancia(Recta3D Recta, Plano3D Plano)
		{
			if (Recta.VectorDirector * Plano.VectorNormal == 0) {
				return Distancia(Recta, Plano.ObtenerPunto(0, 0));
			} else {
				return 0;
			}
		}

		public static double Distancia(Recta3D R1, Recta3D R2)
		{
			switch (PosicionRelativa(R1, R2).Tipo) {
				case TipoPosicionRelativa3D.Coincidente:
					return 0;
				case TipoPosicionRelativa3D.Secante:
					return 0;
				case TipoPosicionRelativa3D.Paralelo:
					return Distancia(R1, R2.ObtenerPuntoParametrico(0));
				case TipoPosicionRelativa3D.Cruce:
					return Math.Abs(Vector3D.ProductoMixto(R1.VectorDirector, R2.VectorDirector, new Vector3D(R1.ObtenerPuntoParametrico(0), R2.ObtenerPuntoParametrico(0)))) / (R1.VectorDirector + R2.VectorDirector).Modulo;
			}
		}

		public static Punto3D Proyeccion(Punto3D Punto, Recta3D Recta)
		{
			int Landa = 0;
			double Vx = 0;
			double Vy = 0;
			double Vz = 0;
			double VVx = 0;
			double VVy = 0;
			double VVz = 0;
			double Qx = 0;
			double Qy = 0;
			double Qz = 0;
			double x = 0;
			double y = 0;
			double z = 0;

			Vx = Recta.VectorDirector.X;
			Vy = Recta.VectorDirector.Y;
			Vz = Recta.VectorDirector.Z;

			VVx = Vx * Vx;
			VVy = Vy * Vy;
			VVz = Vz * Vz;

			Qx = Recta.PuntoInicial.X;
			Qy = Recta.PuntoInicial.Y;
			Qz = Recta.PuntoInicial.Z;

			x = Punto.X;
			y = Punto.Y;
			z = Punto.Z;

			Landa = ((Vx * (x - Qx)) + (Vy * (y - Qy)) + (Vz * (z - Qz))) / (VVx + VVy + VVz);

			return Recta.ObtenerPuntoParametrico(Landa);
		}
	}
}

File: motor3d 2.0/recta3d.vb
Imports Motor3D.Algebra
Imports System.Math

Namespace Espacio3D
    Public Class Recta3D
        Inherits ObjetoGeometrico3D

        Private PlanoA, PlanoB As Plano3D
        Private mVector As Vector3D
        Private mPunto As Punto3D
        Private mPuntoMira As Punto3D

        Public ReadOnly Property PrimerPlano() As Plano3D
            Get
                Return PlanoA
            End Get
        End Property

        Public ReadOnly Property SegundoPlano() As Plano3D
            Get
                Return PlanoB
            End Get
        End Property

        Public ReadOnly Property VectorDirector() As Vector3D
            Get
                Return mVector
            End Get
        End Property

        Public ReadOnly Property PuntoInicial As Punto3D
            Get
                Return mPunto
            End Get
        End Property

        Public ReadOnly Property PuntoMira As Punto3D
            Get
                Return mPuntoMira
            End Get
        End Property

        Public ReadOnly Property Matrices As Matriz()
            Get
                Return RepresentacionMatricial(Me)
            End Get
        End Property

        Public Function ObtenerPuntoParametrico(ByVal Delta As Double) As Punto3D
            Return New Punto3D(mPunto.X + (Delta * mVector.X), mPunto.Y + (Delta * mVector.Y), mPunto.Z + (Delta * mVector.Z))
        End Function

        Public Function ObtenerParametro(ByVal Valor As Double, ByVal Coordenada As EnumEjes) As Double
            Select Case Coordenada
                Case EnumEjes.EjeX
                    If mVector.X <> 0 Then
                        Return (Valor - mPunto.X) / mVector.X
                    Else
                        Throw New ExcepcionGeometrica3D("RECTA3D (OBTENERPARAMETRO): Division por cero." & vbNewLine & _
                                                        "Coordenada: " & Coordenada.ToString & vbNewLine & _
                                                        "Vector direccion: " & mVector.ToString)
                    End If
                Case EnumEjes.EjeY
                    If mVector.Y <> 0 Then
                        Return (Valor - mPunto.Y) / mVector.Y
                    Else
                        Throw New ExcepcionGeometrica3D("RECTA3D (OBTENERPARAMETRO): Division por cero." & vbNewLine & _
                                                        "Coordenada: " & Coordenada.ToString & vbNewLine & _
                                                        "Vector direccion: " & mVector.ToString)
                    End If
                Case EnumEjes.EjeZ
                    If mVector.Z <> 0 Then
                        Return (Valor - mPunto.Z) / mVector.Z
                    Else
                        Throw New ExcepcionGeometrica3D("RECTA3D (OBTENERPARAMETRO): Division por cero." & vbNewLine & _
                                                        "Coordenada: " & Coordenada.ToString & vbNewLine & _
                                                        "Vector direccion: " & mVector.ToString)
                    End If
            End Select
        End Function

        Public Function ObtenerPunto(ByVal Valor As Double, ByVal Coordenada As EnumEjes) As Punto3D
            Return ObtenerPuntoParametrico(ObtenerParametro(Valor, Coordenada))
        End Function

        Public Function ObtenerPunto(ByVal X As Double) As Punto3D
            Dim ec1, ec2 As Ecuacion
            Dim sis As SistemaEcuaciones

            ec1 = New Ecuacion(PlanoA.B, PlanoA.C, -(PlanoA.D + (PlanoA.A * X)))
            ec2 = New Ecuacion(PlanoB.B, PlanoB.C, -(PlanoB.D + (PlanoB.A * X)))

            sis = New SistemaEcuaciones(ec1, ec2)

            If sis.Solucion.TipoSolucion = TipoSolucionSistema.SistemaCompatibleDeterminado Then
                Return New Punto3D(X, sis.Solucion.ValorSolucion(0), sis.Solucion.ValorSolucion(1))
            Else
                Throw New ExcepcionGeometrica3D("RECTA3D (OBTENERPUNTO): No se ha podido calcular el punto. Es posible que los datos de los planos sean erroneos, o que el cálculo del sistema halla fallado." & vbNewLine _
                                                & "Valor de la variable=" & X.ToString & vbNewLine _
                                                & "Primer plano: " & PlanoA.ToString & vbNewLine _
                                                & "Seundo plano: " & PlanoB.ToString & vbNewLine _
                                                & "Primera ecuación del sistema: " & ec1.ToString & vbNewLine _
                                                & "Segunda ecuación del sistema: " & ec2.ToString & vbNewLine _
                                                & "Solución obtenida: " & sis.Solucion.ToString)
            End If
        End Function

        Public Sub New(ByVal P1 As Plano3D, ByVal P2 As Plano3D)
            PlanoA = P1
            PlanoB = P2
            mVector = (P1.VectorNormal & P2.VectorNormal).VectorUnitario
            mPunto = ObtenerPuntoParametrico(0)
            mPuntoMira = ObtenerPuntoParametrico(10)
        End Sub

        Public Sub New(ByVal P1 As Punto3D, ByVal P2 As Punto3D)
            Me.New(P1, New Vector3D(P1, P2))
        End Sub

        Public Sub New(ParamArray Matrices() As Matriz)
            If Matrices.GetUpperBound(0) = 1 Then
                Dim Punto As New Punto3D(Matrices(0))
                Dim Punto2 As New Punto3D(Matrices(1))

                mPunto = Punto
                mPuntoMira = Punto2
                mVector = New Vector3D(mPunto, mPuntoMira).VectorUnitario

                'ÉSTOS VALORES SE OBTIENEN AL DESPEJAR LA ECUACIÓN PARAMÉTRICA DE UNA RECTA GENÉRICA:
                PlanoA = New Plano3D(mVector.Y, -mVector.X, 0, (-mVector.Y * Punto.X) + (mVector.X * Punto.Y))
                PlanoB = New Plano3D(0, mVector.Z, -mVector.Y, (-mVector.Z * Punto.Y) + (mVector.Y * Punto.Z))
            Else
                Throw New ExcepcionGeometrica3D("RECTA3D (NEW): La representación matricial de una recta corresponde a un array con dos matrices" & vbNewLine & _
                                                "Tamaño del array=" & Matrices.GetUpperBound(0) + 1)
            End If
        End Sub

        Public Sub New(ByVal Punto As Punto3D, ByVal Vector As Vector3D)
            mPunto = Punto
            mPuntoMira = New Punto3D(mPunto.X + Vector.X, mPunto.Y + Vector.Y, mPunto.Z + Vector.Z)
            mVector = Vector.VectorUnitario

            'ÉSTOS VALORES SE OBTIENEN AL DESPEJAR LA ECUACIÓN PARAMÉTRICA DE UNA RECTA GENÉRICA:
            PlanoA = New Plano3D(Vector.Y, -Vector.X, 0, (-Vector.Y * Punto.X) + (Vector.X * Punto.Y))
            PlanoB = New Plano3D(0, Vector.Z, -Vector.Y, (-Vector.Z * Punto.Y) + (Vector.Y * Punto.Z))
        End Sub

        Public Function Pertenece(ByVal Punto As Punto3D) As Boolean
            Return PlanoA.Pertenece(Punto) AndAlso PlanoB.Pertenece(Punto)
        End Function

        Public Overrides Function ToString() As String
            Return "{Recta " & PlanoA.ToString & " & " & PlanoB.ToString & "}"
        End Function

        Public Shared Function RepresentacionMatricial(ByVal Recta As Recta3D) As Matriz()
            Dim Retorno(1) As Matriz

            Retorno(0) = Recta.PuntoInicial.Matriz
            Retorno(1) = Recta.PuntoMira.Matriz

            Return Retorno
        End Function

        Public Shared Function PosicionRelativa(ByVal R1 As Recta3D, ByVal R2 As Recta3D) As PosicionRelativa3D
            Dim Sistema As New SistemaEcuaciones(New Ecuacion(R1.VectorDirector.X, R2.VectorDirector.X, R2.PuntoInicial.X - R1.PuntoInicial.Y), _
                                                 New Ecuacion(R1.VectorDirector.Y, R2.VectorDirector.Y, R2.PuntoInicial.Y - R1.PuntoInicial.Y), _
                                                 New Ecuacion(R1.VectorDirector.Z, R2.VectorDirector.Z, R2.PuntoInicial.Z - R1.PuntoInicial.Z))

            Select Case Sistema.Solucion.TipoSolucion
                Case TipoSolucionSistema.SistemaCompatibleDeterminado
                    Return New PosicionRelativa3D(New Punto3D(Sistema.Solucion.ValorSolucion(0), Sistema.Solucion.ValorSolucion(1), Sistema.Solucion.ValorSolucion(2)))
                Case TipoSolucionSistema.SistemaCompatibleIndeterminado
                    Return New PosicionRelativa3D(TipoPosicionRelativa3D.Coincidente)
                Case TipoSolucionSistema.SistemaIncompatible
                    If Sistema.RangoMatrizPrincipal = 1 Then
                        Return New PosicionRelativa3D(TipoPosicionRelativa3D.Paralelo)
                    Else
                        Return New PosicionRelativa3D(TipoPosicionRelativa3D.Cruce)
                    End If
            End Select
        End Function

        Public Shared Function Distancia(ByVal Recta As Recta3D, ByVal Punto As Punto3D) As Double
            Return (New Vector3D(Recta.ObtenerPuntoParametrico(0), Punto) & Recta.VectorDirector).Modulo
        End Function

        Public Shared Function Distancia(ByVal Recta As Recta3D, ByVal Plano As Plano3D) As Double
            If Recta.VectorDirector * Plano.VectorNormal = 0 Then
                Return Distancia(Recta, Plano.ObtenerPunto(0, 0))
            Else
                Return 0
            End If
        End Function

        Public Shared Function Distancia(ByVal R1 As Recta3D, ByVal R2 As Recta3D) As Double
            Select Case PosicionRelativa(R1, R2).Tipo
                Case TipoPosicionRelativa3D.Coincidente
                    Return 0
                Case TipoPosicionRelativa3D.Secante
                    Return 0
                Case TipoPosicionRelativa3D.Paralelo
                    Return Distancia(R1, R2.ObtenerPuntoParametrico(0))
                Case TipoPosicionRelativa3D.Cruce
                    Return Abs(Vector3D.ProductoMixto(R1.VectorDirector, R2.VectorDirector, New Vector3D(R1.ObtenerPuntoParametrico(0), R2.ObtenerPuntoParametrico(0)))) / (R1.VectorDirector & R2.VectorDirector).Modulo
            End Select
        End Function

        Public Shared Function Proyeccion(ByVal Punto As Punto3D, ByVal Recta As Recta3D) As Punto3D
            Dim Landa As Integer
            Dim Vx, Vy, Vz, VVx, VVy, VVz, Qx, Qy, Qz, x, y, z As Double

            Vx = Recta.VectorDirector.X
            Vy = Recta.VectorDirector.Y
            Vz = Recta.VectorDirector.Z

            VVx = Vx * Vx
            VVy = Vy * Vy
            VVz = Vz * Vz

            Qx = Recta.PuntoInicial.X
            Qy = Recta.PuntoInicial.Y
            Qz = Recta.PuntoInicial.Z

            x = Punto.X
            y = Punto.Y
            z = Punto.Z

            Landa = ((Vx * (x - Qx)) + (Vy * (y - Qy)) + (Vz * (z - Qz))) / (VVx + VVy + VVz)

            Return Recta.ObtenerPuntoParametrico(Landa)
        End Function
    End Class
End Namespace

